File: IEEEXtreme13/InfString/J.cpp
#include <iostream>
#include <cmath>

using namespace std;


typedef long long bigint;

char solve(bigint b, bigint x) {
    bigint extra = 1;
    bigint current = 0;
    bigint total = 0;
    int iterations = 0;
    while (total <= x) {
        iterations++;
        extra *= b;
        current = b * current + extra;
        total += current;
    }
    char strings[iterations];
    bigint total_negative = 0;
    bigint mult = 1;
    for (bigint y=0; y<iterations; y++) {
        mult *= b;
        int h = (total - total_negative - x) / (current / mult) - 1;
        bigint negative = ((bigint)h+1) * (bigint)(current / mult);
        if (total - (total_negative + negative) <= x) h--;
        total_negative += ((bigint)h+1) * (bigint)(current / mult);
        strings[y] = ((char)('a' + (int)(b - h - 2)));
    }
    return strings[(x - (total - total_negative)) + iterations];
}

int main() {

    int t;
    cin >> t;

    for (int i=0; i<t; i++) {
        bigint b, x;
        cin >> b >> x;
        cout << solve(b, x) << endl;
    }

    return 0;
}

File: IEEEXtreme13/InfString/J.py
t = int(input())

def solve(b, x):
    extra = 1
    current = 0
    total = 0
    iterations = 0
    while total <= x:
        iterations += 1
        extra *= b
        current = b * current + extra
        total += current
    # Now binary search
    string = ''
    total_negative = 0
    for y in range(iterations):
        negative = 0
        for z in range(b):
            negative = int((z+1) * current / pow(b, y+1))
            if total - total_negative - negative <= x:
                total_negative += int((z) * current / pow(b, y+1))
                string = string + chr(ord('a') + (b-z-1))
                break
    return string[int(x - (total - total_negative))]

for case in range(t):
    print(solve(*map(int, input().split())))
